pkg/operator/controller/gatewayapi/controller_test.go

```diff
@@ -0,0 +1,362 @@
+			expectDelete:    []client.Object{},
+			expectStartCtrl: false,
+		},
+		{
+			name: "featuregates.config.openshift.io/bogus enables GatewayAPI",
+			existingObjects: []runtime.Object{
+				&configv1.FeatureGate{
+					ObjectMeta: metav1.ObjectMeta{Name: "bogus"},
+					Spec: configv1.FeatureGateSpec{
+						FeatureGateSelection: configv1.FeatureGateSelection{
+							FeatureSet: configv1.CustomNoUpgrade,
+							CustomNoUpgrade: &configv1.CustomFeatureGates{
+								Enabled: []string{"GatewayAPI"},
+							},
+						},
+					},
+				},
+			},
+			expectCreate:    []client.Object{},
+			expectUpdate:    []client.Object{},
+			expectDelete:    []client.Object{},
+			expectStartCtrl: false,
+		},
+	}
+
+	scheme := runtime.NewScheme()
+	configv1.AddToScheme(scheme)
+	apiextensionsv1.AddToScheme(scheme)
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			fakeClient := fake.NewClientBuilder().
+				WithScheme(scheme).
+				WithRuntimeObjects(tc.existingObjects...).
+				Build()
+			cl := &fakeClientRecorder{fakeClient, t, []client.Object{}, []client.Object{}, []client.Object{}}
+			informer := informertest.FakeInformers{Scheme: scheme}
+			cache := fakeCache{Informers: &informer, Reader: cl}
+			ctrl := &fakeController{t, false, nil}
+			reconciler := &reconciler{
+				cache:  cache,
+				client: cl,
+				config: Config{
+					DependentControllers: []controller.Controller{ctrl},
+				},
+			}
+			req := reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name: "cluster",
+				},
+			}
+			res, err := reconciler.Reconcile(context.Background(), req)
+			assert.NoError(t, err)
+			assert.Equal(t, reconcile.Result{}, res)
+			ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+			defer cancel()
+			select {
+			case <-ctrl.startNotificationChan:
+				t.Log("Start() was called")
+			case <-ctx.Done():
+				t.Log(ctx.Err())
+			}
+			assert.Equal(t, ctrl.started, tc.expectStartCtrl, "fake controller should have been started")
+			cmpOpts := []cmp.Option{
+				cmpopts.EquateEmpty(),
+				cmpopts.IgnoreFields(metav1.ObjectMeta{}, "Annotations", "ResourceVersion"),
+				cmpopts.IgnoreFields(metav1.TypeMeta{}, "Kind", "APIVersion"),
+				cmpopts.IgnoreFields(apiextensionsv1.CustomResourceDefinition{}, "Spec"),
+			}
+			if diff := cmp.Diff(tc.expectCreate, cl.added, cmpOpts...); diff != "" {
+				t.Fatalf("found diff between expected and actual creates: %s", diff)
+			}
+			if diff := cmp.Diff(tc.expectUpdate, cl.updated, cmpOpts...); diff != "" {
+				t.Fatalf("found diff between expected and actual updates: %s", diff)
+			}
+			if diff := cmp.Diff(tc.expectDelete, cl.deleted, cmpOpts...); diff != "" {
+				t.Fatalf("found diff between expected and actual deletes: %s", diff)
+			}
+		})
+	}
+}
+
+func TestReconcileOnlyStartsControllerOnce(t *testing.T) {
+	scheme := runtime.NewScheme()
+	configv1.AddToScheme(scheme)
+	apiextensionsv1.AddToScheme(scheme)
+	fg := &configv1.FeatureGate{
```

Comment:
You mean so that `TestReconcileOnlyStartsControllerOnce` and the "featuregates.config.openshift.io/cluster specifies CustomNoUpgrade and enables GatewayAPI" test case in `Test_Reconcile` could use the same definition?  I could, but I don't think that would really make the tests easier to read, especially since the definition is only used in two places (in separate tests, so there'd be a bit of jumping around, to this global definition and then back to the test) and all the other test cases in `Test_Reconcile` use different featuregate definitions (so the odd one out would be a bit jarring when reading through the test cases).

Reply:
Ack, makes sense.